// videogular2/compiled/src/controls/vg-playback-button/vg-playback-button.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var vg_playback_button_1 = require("./vg-playback-button");
var vg_api_1 = require("../../core/services/vg-api");
var vg_states_1 = require("../../core/states/vg-states");
describe('Playback Button', function () {
    var playbackButton;
    var ref;
    var api;
    beforeEach(function () {
        ref = {
            nativeElement: {
                getAttribute: function (name) {
                    return name;
                }
            }
        };
        api = new vg_api_1.VgAPI();
        api.medias = {
            main: {
                state: vg_states_1.VgStates.VG_PLAYING
            },
            secondary: {
                state: vg_states_1.VgStates.VG_PAUSED
            }
        };
        playbackButton = new vg_playback_button_1.VgPlaybackButton(ref, api);
    });
    it('Should set playbackIndex default value to 1', function () {
        expect(playbackButton.playbackIndex).toEqual(1);
    });
    it('Should get media by id on init', function () {
        spyOn(api, 'getMediaById').and.callFake(function () { return ({}); });
        playbackButton.vgFor = 'test';
        playbackButton.onPlayerReady();
        expect(api.getMediaById).toHaveBeenCalledWith('test');
    });
    describe('onClick (single and multiple media)', function () {
        it('should increase playbackIndex', function () {
            api.medias = {
                main: {
                    state: vg_states_1.VgStates.VG_PLAYING
                }
            };
            playbackButton.target = api;
            playbackButton.onClick();
            expect(playbackButton.playbackIndex).toEqual(2);
        });
        it('should set playbackRate to target media', function () {
            api.medias = {
                main: {
                    state: vg_states_1.VgStates.VG_PLAYING
                }
            };
            playbackButton.target = api;
            playbackButton.onClick();
            expect(playbackButton.target.playbackRate).toEqual('1.5');
        });
        it('should set playbackRate to target media', function () {
            var media = {
                playbackRate: {
                    test: '1'
                }
            };
            playbackButton.target = media;
            playbackButton.vgFor = 'test';
            playbackButton.onClick();
            expect(playbackButton.target.playbackRate.test).toEqual('1.5');
        });
    });
});
