// videogular2/compiled/src/buffering/vg-buffering.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var vg_buffering_1 = require("./vg-buffering");
var vg_api_1 = require("../core/services/vg-api");
describe('Buffering', function () {
    var vgBuffering;
    var ref;
    var api;
    beforeEach(function () {
        ref = {
            nativeElement: {
                getAttribute: function (name) {
                    return name;
                }
            }
        };
        api = new vg_api_1.VgAPI();
        vgBuffering = new vg_buffering_1.VgBuffering(ref, api);
    });
    describe('onPlayerReady', function () {
        it('should subscribe to bufferDetected media events', function () {
            spyOn(api, 'getMediaById').and.returnValue({
                subscriptions: {
                    bufferDetected: { subscribe: jasmine.createSpy('bufferDetected') }
                }
            });
            vgBuffering.onPlayerReady();
            expect(vgBuffering.target.subscriptions.bufferDetected.subscribe).toHaveBeenCalled();
        });
    });
    describe('isBuffering', function () {
        it('should show if buffer is detected', function () {
            vgBuffering.onUpdateBuffer(true);
            expect(vgBuffering.isBuffering).toBe(true);
        });
        it('should hide if buffer is not detected', function () {
            vgBuffering.onUpdateBuffer(false);
            expect(vgBuffering.isBuffering).toBe(false);
        });
    });
});
