// videogular2/compiled/src/controls/vg-scrub-bar/vg-scrub-bar-current-time/vg-scrub-bar-current-time.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var vg_scrub_bar_current_time_1 = require("./vg-scrub-bar-current-time");
var vg_api_1 = require("../../../core/services/vg-api");
describe('Scrub bar current time', function () {
    var scrubBarCurrentTime;
    var ref;
    var api;
    beforeEach(function () {
        ref = {
            nativeElement: {
                getAttribute: function (name) {
                    return name;
                }
            }
        };
        api = new vg_api_1.VgAPI();
        scrubBarCurrentTime = new vg_scrub_bar_current_time_1.VgScrubBarCurrentTime(ref, api);
    });
    it('Should get media by id on init', function () {
        spyOn(api, 'getMediaById');
        scrubBarCurrentTime.vgFor = 'test';
        scrubBarCurrentTime.onPlayerReady();
        expect(api.getMediaById).toHaveBeenCalledWith('test');
    });
    describe('getPercentage', function () {
        it('should return 50% when current time is 10 and total time is 20', function () {
            scrubBarCurrentTime.target = {
                time: {
                    current: 10,
                    total: 20
                }
            };
            var percent = scrubBarCurrentTime.getPercentage();
            expect(percent).toEqual('50%');
        });
        it('should return 25% when current time is 5 and total time is 20', function () {
            scrubBarCurrentTime.target = {
                time: {
                    current: 5,
                    total: 20
                }
            };
            var percent = scrubBarCurrentTime.getPercentage();
            expect(percent).toEqual('25%');
        });
    });
});
